File: Hard/2026-01-14-3454-Separate-Squares-II/solution.py
class Solution:
    def minY(self, squares: List[List[int]]) -> float:
        events = []
        for x, y, l in squares:
            events.append((x, y, y + l, 1))  # Start of square
            events.append((x + l, y, y + l, -1))  # End of square
        
        events.sort()
        
        active_intervals = []
        total_area = 0
        
        def calculate_area():
            if not active_intervals:
                return 0
            total = 0
            current_start = -1
            current_end = -1
            for start, end in active_intervals:
                if current_end < start:
                    total += current_end - current_start
                    current_start = start
                    current_end = end
                else:
                    current_end = max(current_end, end)
            total += current_end - current_start
            return total
        
        last_x = events[0][0]
        for x, y1, y2, typ in events:
            total_area += calculate_area() * (x - last_x)
            if typ == 1:
                active_intervals.append((y1, y2))
                active_intervals.sort()
            else:
                active_intervals.remove((y1, y2))
            last_x = x
        
        half_area = total_area / 2
        current_area = 0
        last_y = 0
        
        for y in sorted(set(y for _, y1, y2, _ in events for y in (y1, y2))):
            current_area += calculate_area() * (y - last_y)
            if current_area >= half_area:
                return float(y)
            last_y = y
        
        return -1.0

File: Hard/2026-01-14-3454-Separate-Squares-II/solution.js
class Solution {
    separateSquares(squares) {
        const events = [];
        const areaMap = new Map();

        for (const [x, y, l] of squares) {
            events.push([x, y, y + l, 1]);
            events.push([x + l, y, y + l, -1]);
        }

        events.sort((a, b) => a[0] - b[0] || a[3] - b[3]);

        let totalArea = 0;
        let activeIntervals = [];

        const updateArea = () => {
            let currentArea = 0;
            let lastY = -1;

            for (const [start, end] of activeIntervals) {
                if (lastY < start) {
                    currentArea += end - start;
                    lastY = end;
                } else if (lastY < end) {
                    currentArea += end - lastY;
                    lastY = end;
                }
            }

            return currentArea;
        };

        for (const [x, startY, endY, type] of events) {
            if (type === 1) {
                activeIntervals.push([startY, endY]);
                activeIntervals.sort((a, b) => a[0] - b[0]);
            } else {
                activeIntervals = activeIntervals.filter(interval => !(interval[0] === startY && interval[1] === endY));
            }

            totalArea = updateArea();
        }

        let halfArea = totalArea / 2;
        let currentArea = 0;
        let result = 0;

        for (const [x, startY, endY, type] of events) {
            if (type === 1) {
                activeIntervals.push([startY, endY]);
                activeIntervals.sort((a, b) => a[0] - b[0]);
            } else {
                activeIntervals = activeIntervals.filter(interval => !(interval[0] === startY && interval[1] === endY));
            }

            currentArea = updateArea();

            if (currentArea >= halfArea) {
                result = (startY + endY) / 2;
                break;
            }
        }

        return result;
    }
}